src/core/lib/surface/call.c

```diff
@@ -1290,11 +1303,11 @@ static void receiving_stream_ready(grpc_exec_ctx *exec_ctx, void *bctlp,
     cancel_with_error(exec_ctx, call, STATUS_FROM_SURFACE,
                       GRPC_ERROR_REF(error));
   }
-  if (call->has_initial_md_been_received || error != GRPC_ERROR_NONE ||
-      call->receiving_stream == NULL) {
+  call->saved_receiving_stream_ready_bctlp = bctlp;
```

Comment:
We talked offline about this yesterday but putting it here again for the record.   This change to `saved_receiving_stream_ready_bctlp` may not be visible after the `gpr_atm_acq_cas` you are doing in `receiving_initial_metadata_ready` function. (please see my comment there)